Add mutated_args field to custom_op

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123454
* #123453
* #123437
* __->__ #123129
* #123110
* #123109
* #123108

If provided, we:
- autogenerate an ADInplaceOrView implementation
- assume that no mutated inputs are returned as outputs. There are
  already aliasing runtime checks that check this.

Test Plan:
- new tests

zou3519 committed:
Add mutated_args field to @custom_op

If provided, we:
- autogenerate an ADInplaceOrView implementation
- assume that no mutated inputs are returned as outputs. There are
  already aliasing runtime checks that check this.

Test Plan:
- new tests

[ghstack-poisoned]

albanD commented on test/test_custom_ops.py:
```diff
@@ -2117,6 +2117,47 @@ def _(x, y):
 
+    def test_mutated(self):
+        @torch.library.custom_op(
+            "_torch_testing::numpy_sin_inplace", mutated_args={"x"}, device_types="cpu"
```
Why cpu only here?

zou3519 replied:
otherwise np.sin(x_np, out=x_np) doesn't actually update the original input

albanD commented on test/test_custom_ops.py:
```diff
@@ -2117,6 +2117,47 @@ def _(x, y):
+        self.assertEqual(x, expected)
+        self.assertGreater(x._version, version)
+
+        @torch.library.custom_op("_torch_testing::f", mutated_args="yzwv")
```
I don't think this should be allowed! This is not going to work for any non-trivially named arguments and is very confusing to read.

zou3519 replied:
Sure

zou3519 committed:
Update on "Add mutated_args field to @custom_op"


If provided, we:
- autogenerate an ADInplaceOrView implementation
- assume that no mutated inputs are returned as outputs. There are
  already aliasing runtime checks that check this.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add mutated_args field to @custom_op"


If provided, we:
- autogenerate an ADInplaceOrView implementation
- assume that no mutated inputs are returned as outputs. There are
  already aliasing runtime checks that check this.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add mutated_args field to custom_op"


If provided, we:
- autogenerate an ADInplaceOrView implementation
- assume that no mutated inputs are returned as outputs. There are
  already aliasing runtime checks that check this.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add mutated_args field to custom_op"


If provided, we:
- autogenerate an ADInplaceOrView implementation
- assume that no mutated inputs are returned as outputs. There are
  already aliasing runtime checks that check this.

Test Plan:
- new tests

[ghstack-poisoned]

albanD commented on torch/_custom_op/impl.py:
```diff
@@ -811,7 +812,14 @@ def error_fn(what):
             schema_type = f"Tensor(a{idx}!){schema_type[len('Tensor'):]}"
+        seen_args.add(name)
         params.append(f"{schema_type} {name}")
+    mutated_args_not_seen = set(mutated_args) - seen_args
```
Shouldn't mutated_args always be a set?

zou3519 replied:
It's an Iterable (list, tuple, set, etc). I didn't want to make it set all the time because the empty set is kinda weird: set()

albanD commented on torch/_library/custom_ops.py:
```diff
@@ -378,6 +399,26 @@ def fake_impl(*args, **kwargs):
+                        continue
+                    if isinstance(val, Tensor):
+                        autograd.graph.increment_version(val)
+                    elif isinstance(val, (tuple, list)):
```
Don't we ban TensorList?

zou3519 replied:
TensorList banned in autograd but not for other custom ops.

albanD reviewed (approved):
Sounds good

zou3519 committed:
Update on "Add mutated_args field to custom_op"


If provided, we:
- autogenerate an ADInplaceOrView implementation
- assume that no mutated inputs are returned as outputs. There are
  already aliasing runtime checks that check this.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add mutated_args field to custom_op"


If provided, we:
- autogenerate an ADInplaceOrView implementation
- assume that no mutated inputs are returned as outputs. There are
  already aliasing runtime checks that check this.

Test Plan:
- new tests

[ghstack-poisoned]
